logger/factory.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package logger
+
+import (
+	"fmt"
+	"github.com/uber-common/bark"
+)
+
+// LoggerFactory wraps a bark.Logger and provides a way to create named loggers
+// restricted to a specific Level.
+type LoggerFactory struct {
+	logger bark.Logger
+	cache  map[string]*namedLogger
+}
+
+func NewLoggerFactory(l bark.Logger) *LoggerFactory {
+	return &LoggerFactory{
+		logger: l,
+		cache:  make(map[string]*namedLogger),
+	}
+}
+
+// SetLevel sets the minimum level for a named logger. A named logger emits
+// messages with a level equal to or greater than this level.
+func (lf *LoggerFactory) SetLevel(name string, level Level) error {
```

Comment:
Oh, I see what you're saying. There are a few things to keep in mind:
* The factory interface makes no promises that `factory.Logger("x")` always returns the same instance.
* Is not enough to propagate the level change only on the logger returned by `factory.Logger("x")`. The level change must be propagated to all loggers subsequently created by `WithField`/`WithFields` calls. (i.e. `logger := factory.Logger("x").WithField("key", value)`)
* Once the level propagation fix lands the API will make more sense because the levels are stored in the factory itself and `namedLoggers` instances forward messages back to the factory. This arranges the code in a way that propagation (which turns out is difficult to achieve) is unnecessary.

But I do get your point, and for the current implementation I agree with it.